src/cnfizers/Cnfizer.cc

```diff
@@ -317,178 +137,92 @@ bool Cnfizer::isClause(PTRef e) {
-{
-    Map<PTRef, bool, PTRefHash, Equal<PTRef> > check_cache;
-    const Pterm &not_t = logic.getPterm (e);
-
-    if (logic.isNot (e) && checkPureConj (not_t[0], check_cache) ) return true;
-    else return false;
+// Check if this is a negation of a clause
+bool Cnfizer::checkDeMorgan(PTRef e) {
+    Pterm const & term = logic.getPterm(e);
+    if (not logic.isNot(term.symb())) return false;
+    Map<PTRef, bool, PTRefHash, Equal<PTRef>> check_cache;
+    return checkPureConj(term[0], check_cache);
 }
 
-//
 // Check whether it is a pure conjunction of literals
-//
-bool Cnfizer::checkPureConj (PTRef e, Map<PTRef, bool, PTRefHash, Equal<PTRef> > &check_cache)
-{
-    if (check_cache.has (e))
-        return true;
+bool Cnfizer::checkPureConj(PTRef e, Map<PTRef, bool, PTRefHash, Equal<PTRef>> & check_cache) {
+    if (check_cache.has(e)) return true;
+    if (not logic.isAnd(e)) return false;
 
     vec<PTRef> to_process;
-    to_process.push (e);
-
-    // Topmost needs to be and
-    if (!logic.isAnd (e)) return false;
-
-    while (to_process.size() != 0)
-    {
-        e = to_process.last();
+    to_process.push(e);
+    while (to_process.size() != 0) {
+        PTRef current = to_process.last();
         to_process.pop();
-        Pterm &and_t = logic.getPterm (e);
 
-        if (logic.isAnd(e)) {
-            for (PTRef tr : and_t) {
+        if (logic.isAnd(current)) {
+            for (PTRef tr : logic.getPterm(current)) {
                 to_process.push(tr);
             }
-        } else if (not isLiteral(e)) {
+        } else if (not isLiteral(current)) {
             return false;
         }
-        check_cache.insert (e, true);
+        check_cache.insert(current, true);
     }
-
     return true;
 }
 
-bool Cnfizer::addClause(const vec<Lit> & c_in)
+void Cnfizer::addClause(vec<Lit> && clause)
 {
-    vec<Lit> c;
-    c_in.copyTo(c);
-    if (current_frame_term != logic.getTerm_true()) {
-        Lit l = this->getOrCreateLiteralFor(current_frame_term);
-        Var v = var(l);
-        tmap.setFrozen(v);
-        solver.addAssumptionVar(v);
-        c.push(l);
-    }
-
-#ifdef PEDANTIC_DEBUG
-    cerr << "== clause start" << endl;
-
-    for (int i = 0; i < c.size(); i++)
-    {
-        cerr << "(" << c[i].x << "," << var (c[i]) << ") * " << (sign (c[i]) ? "not " : "")
-             << logic.printTerm (tmap.varToPTRef (var (c[i])))
-             << " ";
-        cerr << endl;
-    }
-
-#endif
-    opensmt::pair<CRef, CRef> iorefs{CRef_Undef, CRef_Undef};
-    bool res = solver.addOriginalSMTClause(c, iorefs);
-    if (keepPartitionInfo()) {
-        CRef ref = iorefs.first;
-        if (ref != CRef_Undef) {
-            ipartitions_t parts = 0;
-            assert(currentPartition != -1);
-            opensmt::setbit(parts, static_cast<unsigned int>(currentPartition));
-            pmanager.addClauseClassMask(ref, parts);
-        }
-    }
-    return res;
+    (*clauseCallBack)(std::move(clause));
```

Comment:
That is a bit weird func, clauseCallBack is not static, I'm not sure how it works tbh
Where does it add a clause now?

Reply:
`clauseCallBack` is a pointer to a function object, we are just calling it here to process the `clause`.

Reply:
But it is never assigned anything. This is the only place `clauseCallBack` pointer is actually used and it is directing to the empty object (I searched by the `clauseCallBack` name)

Reply:
The callback is set using `setClauseCallBack` method and we set the callback in `MainSolver::giveToSolver`.

Reply:
Aaaaah, sry, I missed the fact that you have "recreated" the same struct inside a function

Now it seems solid!